lookout/style/typos/corrector.py

```diff
@@ -77,118 +82,88 @@ def initialize_generator(self, vocabulary_file: str, frequencies_file: str,
-                     return_all: bool = True) -> Dict[int, List[Tuple[str, float]]]:
-        """
-        Suggest corrections for given typos.
-
-        :param typos_file: csv file containing DataFrame with column "typo", \
-                           column "token_split" is optional, but used when present
-        :param candidates_file: pickle file containing DataFrame with precalculated \
-                                candidates and features
-        :param n_candidates: Number of most probable candidates to return
-        :param return_all: False to return suggestions only for corrected tokens
-        :param save_candidates_file: Path to file to save candidates to
-        :return: Dictionary {id : [[candidate, correctness_proba]]}, candidates are sorted \
-                 by correctness probability in a descending order.
-        """
-        typos = pandas.read_csv(typos_file, index_col=0)
-        candidates = None
-        if candidates_file is not None:
-            candidates = pandas.read_pickle(candidates_file)
-        return self.suggest(typos, candidates, save_candidates_file, n_candidates, return_all)
-
-    def suggest_by_batches(self, typos: pandas.DataFrame, n_candidates: int = None,
+    def suggest_by_batches(self, data: Union[pandas.DataFrame, str], n_candidates: int = None,
                            return_all: bool = True, batch_size: int = 2048,
                            ) -> Dict[int, List[Tuple[str, float]]]:
         """
         Correct typos from dataset by batches. Does not support precalculated candidates.
 
         Suggest corrections for given typos
-        :param typos: DataFrame containing column "typo", \
-               column "token_split" is optional, but used when present
+        :param data: DataFrame or its .csv dump, containing column "typo", \
+                     column "token_split" is optional, but used when present
         :param n_candidates: Number of most probable candidates to return
         :param return_all: False to return suggestions only for corrected tokens
         :param batch_size: Batch size
         :return: Dictionary {id : [[candidate, correctness_proba]]}, candidates are sorted \
                  by correctness probability in a descending order.
         """
+        if isinstance(data, str):
+            data = pandas.read_csv(data, index_col=0)
         all_suggestions = []
-        for i in tqdm(range(0, len(typos), batch_size)):
-            suggestions = self.suggest(typos.loc[typos.index[i]:
-                                                 typos.index[min(len(typos) - 1,
-                                                                 i + batch_size - 1)], :],
-                                       n_candidates=n_candidates, return_all=return_all)
+        for i in tqdm(range(0, len(data), batch_size)):
+            suggestions = self.suggest(
+                data.loc[data.index[i]:data.index[min(len(data) - 1, i + batch_size - 1)], :],
```

Comment:
maybe 
```python
data.loc[data.index[i]:data.index[min(len(data), i + batch_size)], :],
```
For examnple
```python
l = [0, 1, 2, 3, 4, 5, 6]
l[1:4]
```
gives you `[1, 2, 3]` without `4`

Reply:
Because apparantly pandas.DataFrame.loc includes both ends (I was surprised too):

> Slice with integer labels for rows. As mentioned above, note that both the start and stop of the slice are included.

from https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.loc.html

Reply:
Wow this is unexpected at all

Reply:
Woooooow

Reply:
can you place a comment with the explanation that it, as is, should be and a link to the doc, please?

Reply:
Ok

Reply:
Maybe you need `iloc`? `loc` is label based. With `iloc` I'd write

    data.iloc[i:i + batch_size, :]
I didn't read the rest of the code so there is maybe a good reason to use `loc` I'm mentioning it just in case.

Reply:
Yes, it's much better this way, thank you!